[DataLoader] Add `generate_state` for NumPy seeding

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56797 [DataLoader] Add `generate_state` for NumPy seeding**

After adding default seeding strategy for NumPy random module within each worker of DataLoader #56488, two concerns are raised:
- We dropped the support for NumPy < 1.17 due to `SeedSequence`
- In order to support seeding for NumPy < 1.17, how can we provide seed for `numpy.random`?
  - First option is set the same seed as `random`. But, the problem is a same algorithm is shared between `numpy.random` and `random`. With the same seed, they will have exact same state sequence. Thanks to @rkern, we noticed this so-called [bad things](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659).
  - Considering most of users do not aware this problem, we can provide a better seed by default for `numpy.random` using same `SeedSequence` algorithm as numpy. This is just a workaround with hard-coded function to generate an array of four int32 as the seed.

To better coping with this problem since there are amount of 3rd party libraries not just `NumPy` having random module. We may at the end need to implement a `SeedSequence` within `torch.random` module, then users can `spawn` a new `SeedSequence` for each library.

Differential Revision: [D28000619](https://our.internmc.facebook.com/intern/diff/D28000619)

ejguan committed:
[DataLoader] Add `generate_state` for NumPy seeding

[ghstack-poisoned]

rkern commented:
We have some [reference vectors available](https://github.com/numpy/numpy/blob/main/numpy/random/tests/test_seed_sequence.py) in case you want to test for compatibility with numpy's `SeedSequence` and the original C++ implementation.

ejguan committed:
Update on "[DataLoader] Add `generate_state` for NumPy seeding"


After adding default seeding strategy for NumPy random module within each worker of DataLoader #56488, two concerns are raised:
- We dropped the support for NumPy < 1.17 due to `SeedSequence`
- In order to support seeding for NumPy < 1.17, how can we provide seed for `numpy.random`?
  - First option is set the same seed as `random`. But, the problem is a same algorithm is shared between `numpy.random` and `random`. With the same seed, they will have exact same state sequence. Thanks to @rkern, we noticed this so-called [bad things](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659).
  - Considering most of users do not aware this problem, we can provide a better seed by default for `numpy.random` using same `SeedSequence` algorithm as numpy. This is just a workaround with hard-coded function to generate an array of four int32 as the seed.

To better coping with this problem since there are amount of 3rd party libraries not just `NumPy` having random module. We may at the end need to implement a `SeedSequence` within `torch.random` module, then users can `spawn` a new `SeedSequence` for each library.

[ghstack-poisoned]

rkern commented on test/test_dataloader.py:
```diff
@@ -1529,6 +1529,22 @@ def __len__(self):
+            for worker_id in range(8):
+                # Same seedsequence as `_generate_state`
+                ss = [worker_id, bs1, bs2, 0]
+                np_ss = np.random.SeedSequence(ss)
```
This will mean your test suite has a dependency on numpy 1.17+, which is what you were trying to avoid by introducing `_generate_state()`. I recommend just taking the [explicit reference test vectors](https://github.com/numpy/numpy/blob/main/numpy/random/tests/test_seed_sequence.py#L7-L54) that we use in `numpy` to verify the correctness of the implementation.

rkern replied:
Or if you want to test this particular scheme of constructing `ss = [worker_id, ...]`, then just make your own test vectors ahead of time with `SeedSequence` and putting them explicitly here.

ejguan replied:
Oops, thanks for pointing it out. I will generate a list of test cases for reference.

ejguan committed:
Update on "[DataLoader] Add `generate_state` for NumPy seeding"


After adding default seeding strategy for NumPy random module within each worker of DataLoader #56488, two concerns are raised:
- We dropped the support for NumPy < 1.17 due to `SeedSequence`
- In order to support seeding for NumPy < 1.17, how can we provide seed for `numpy.random`?
  - First option is set the same seed as `random`. But, the problem is a same algorithm is shared between `numpy.random` and `random`. With the same seed, they will have exact same state sequence. Thanks to @rkern, we noticed this so-called [bad things](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659).
  - Considering most of users do not aware this problem, we can provide a better seed by default for `numpy.random` using same `SeedSequence` algorithm as numpy. This is just a workaround with hard-coded function to generate an array of four int32 as the seed.

To better coping with this problem since there are amount of 3rd party libraries not just `NumPy` having random module. We may at the end need to implement a `SeedSequence` within `torch.random` module, then users can `spawn` a new `SeedSequence` for each library.

Differential Revision: [D28000619](https://our.internmc.facebook.com/intern/diff/D28000619)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Add `generate_state` for NumPy seeding"


After adding default seeding strategy for NumPy random module within each worker of DataLoader #56488, two concerns are raised:
- We dropped the support for NumPy < 1.17 due to `SeedSequence`
- In order to support seeding for NumPy < 1.17, how can we provide seed for `numpy.random`?
  - First option is set the same seed as `random`. But, the problem is a same algorithm is shared between `numpy.random` and `random`. With the same seed, they will have exact same state sequence. Thanks to @rkern, we noticed this so-called [bad things](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659).
  - Considering most of users do not aware this problem, we can provide a better seed by default for `numpy.random` using same `SeedSequence` algorithm as numpy. This is just a workaround with hard-coded function to generate an array of four int32 as the seed.

To better coping with this problem since there are amount of 3rd party libraries not just `NumPy` having random module. We may at the end need to implement a `SeedSequence` within `torch.random` module, then users can `spawn` a new `SeedSequence` for each library.

Differential Revision: [D28000619](https://our.internmc.facebook.com/intern/diff/D28000619)

[ghstack-poisoned]

rkern commented:
LGTM, for the parts that I am competent to review.
